pkg/render/common/test/testing.go

```diff
@@ -88,7 +90,55 @@ func ExpectResourceInList(objs []client.Object, name, ns, group, version, kind s
 	Expect(elems).To(ContainElement(o))
 }
 
-func ExpectResource(resource runtime.Object, name, ns, group, version, kind string) {
+// ExpectResources checks that the given list of resources contains the expected resources, and that
+// the given resource list does not contain any unexpected resources.
+func ExpectResources(resources []client.Object, expected []client.Object) {
+	// First, check that each actual resource is in the expected list.
+	for _, resource := range resources {
+		ExpectWithOffset(1, ExpectResource(resource, expected)).NotTo(HaveOccurred(), "Unexpected resouce was rendered")
+	}
+
+	// Then, check that each expected resource is in the actual list.
+	for _, resource := range expected {
+		ExpectWithOffset(1, ExpectResource(resource, resources)).NotTo(HaveOccurred(), "Expected resource was not rendered")
+	}
+}
+
+// ExpectResource checks that the given list of resources contains a resource with the given name and
+// namespace, and that the resource has the given GroupVersionKind.
+func ExpectResource(expected client.Object, resources []client.Object) error {
+	name := expected.(metav1.ObjectMetaAccessor).GetObjectMeta().GetName()
+	ns := expected.(metav1.ObjectMetaAccessor).GetObjectMeta().GetNamespace()
+
+	for _, resource := range resources {
+		if reflect.TypeOf(resource) == reflect.TypeOf(expected) {
+			if resource.(metav1.ObjectMetaAccessor).GetObjectMeta().GetName() == name &&
+				resource.(metav1.ObjectMetaAccessor).GetObjectMeta().GetNamespace() == ns {
+				// Same type, name, and namespace. Consider it a match.
+				return nil
+			}
+		}
+	}
+
+	// Build a list of items so we can print a nice error message.
+	items := []string{}
+	for i, r := range resources {
+		items = append(items,
+			fmt.Sprintf("%d: %T Namespace=%s Name=%s",
+				i,
+				r,
+				r.(metav1.ObjectMetaAccessor).GetObjectMeta().GetNamespace(),
+				r.(metav1.ObjectMetaAccessor).GetObjectMeta().GetName(),
+			))
+	}
+	if ns == "" {
+		return fmt.Errorf("%T %s not found in:\n\n%s", expected, name, strings.Join(items, "\n"))
+	}
+	return fmt.Errorf("%T %s/%s not found in:\n\n%s", expected, ns, name, strings.Join(items, "\n"))
+}
+
+// CompareResource checks that the given resource matches the expected name, namespace, group, version, and kind.
+func CompareResource(resource runtime.Object, name, ns, group, version, kind string) {
```

Comment:
I'm not a big fan of this name change. Compare* suggests to me that it would be providing a result. Expect* suggests it will result in an Expect call. Like I think ExpectResourceMeta or something like that would be more informative.

Reply:
Yeah I agree. I'd like to move all the tests off of this function eventually but for now I will change the name to `ExpectResourceTypeAndObjectMeta` or something.